File: src/Skybrud.Umbraco.PackageManifests/Models/Umbraco/PropertyEditorEditorValidation.cs
using System;
using Newtonsoft.Json;

namespace Skybrud.Umbraco.PackageManifests.Models.Umbraco {

    public class PropertyEditorEditorValidation {

        [JsonProperty("mandatory")]
        public bool Mandatory { get; set; }

        [JsonProperty("pattern")]
        public string Pattern { get; set; }

        public bool ShouldSerialize() {
            return ShouldSerializeMandatory() || ShouldSerializePattern();
        }

        public bool ShouldSerializeMandatory() {
            return Mandatory;
        }

        public bool ShouldSerializePattern() {
            return String.IsNullOrWhiteSpace(Pattern) == false;
        }

    }

}

File: src/Skybrud.Umbraco.PackageManifests/App_Plugins/Skybrud.PackageManifests/Scripts/Controllers/Editors/PropertyEditorEditorValidation.js
angular.module("umbraco").controller("PackageManifests.PropertyEditorEditorValidation.Controller", function ($scope, editorService) {

    if (!$scope.model.value) $scope.model.value = {};

    $scope.edit = function() {

        var properties = [
            {
                alias: "mandatory",
                label: "Mandatory",
                description: "",
                hideLabel: false,
                value: $scope.model.value.mandatory ? "1" : "0",
                view: "boolean"
            },
            {
                alias: "pattern",
                label: "Pattern",
                description: "",
                hideLabel: false,
                value: $scope.model.value.pattern,
                view: "textbox"
            }
        ];

        editorService.open({
            title: "Editor Validation",
            view: "/App_Plugins/Skybrud.PackageManifests/Views/Overlays/Properties.html",
            properties: properties,
            submit: function (model) {
                angular.forEach(model, function (p) {
                    if (p.view === "boolean") {
                        $scope.model.value[p.alias] = p.value === "1";
                    } else {
                        $scope.model.value[p.alias] = p.value;
                    }
                });
                editorService.close();
            },
            close: function () {
                editorService.close();
            }
        });

    }

});